Devicetree: Macro to initialize driver for every device instance

io-pa opened the pull request:
This pull request introduces the macro: DT_INST_FOREACH(inst_expr).

This makes it possible to do this:
`DT_INST_FOREACH(I2C_NRFX_TWIM_DEVICE);`

Instead of:
```
#ifdef DT_INST_N_0_nordic_nrf_twim
I2C_NRFX_TWIM_DEVICE(0);
#endif

#ifdef DT_INST_N_1_nordic_nrf_twim
I2C_NRFX_TWIM_DEVICE(1);
#endif

#ifdef DT_INST_N_2_nordic_nrf_twim
I2C_NRFX_TWIM_DEVICE(2);
#endif

#ifdef DT_INST_N_3_nordic_nrf_twim
I2C_NRFX_TWIM_DEVICE(3);
#endif
```


The macro calls the passed inst_expr for every compatible (and enabled) device in the device tree.
By using this approach the drivers don't have to make assumption about the maximum count of device instances anymore.
The expr argument can be either a macro or a function that take the instance number as argument.

io-pa committed:
devicetree.h: Add macro DT_INST_FOREACH

With this macro device drivers can call macros and functions
on every device instance compatible to that driver.
This makes it possible to make drivers agnostic to the
potential counts of instances.
Sidenote: Introduces helper macro DT_CALL_WITH_ARG.

Signed-off-by: Ioannis Papamanoglou <iopapamanoglou@gmail.com>

tbursztyka commented on include/device.h:
```diff
@@ -85,6 +85,25 @@ extern "C" {
 	data, cfg_info, level, prio, NULL)
 
+/**
+ * @def DEVICE_INIT
```
wrong name

io-pa replied:
oh yes thanks!

stephanosio commented on include/device.h:
```diff
@@ -85,6 +85,25 @@ extern "C" {
 	data, cfg_info, level, prio, NULL)
 
+/**
+ * @def DEVICE_INIT
```
This should be `@def DEVICE_INIT_ALL_INSTANCES`

io-pa replied:
oh yes thanks!

tbursztyka commented on include/device.h:
```diff
@@ -85,6 +85,25 @@ extern "C" {
+ * always driver_name. This macro can be used to call the init macro of a driver
+ * for each device specified in the device tree.
+ *
+ * @param init_func Macro that is called for each device node. Has to accept
```
I would name this inst_func instead

io-pa replied:
I changed it. I am not 100% sure though. They macro is called DEVICE_INIT_ALL_INSTANCES so it assumes the passed macro is used to initialize the driver.
In theory we could rename the macro to: DEVICE_CALL_PER_INSTANCE or something like that, since it is not limited to initialization.

tbursztyka replied:
The pb with init_func is that it is also a name used for existing macros. It's called init_fn so it 2 letters different, but it's confusing.

Or cal it init_macro, since your init_func should be macro.

io-pa replied:
I change it to DEVICE_EXEC_FOR_ALL_INSTANCEs and renamed it to inst_macro.
Thanks for the input.

stephanosio reviewed (commented):
Thanks for submitting this. This will be very useful for initialising multiple device instances.

mbolivar-nordic reviewed (commented):
I think ideally this would be rebased on https://github.com/zephyrproject-rtos/zephyr/pull/23245 and added to devicetree.h instead. @galak?

mbolivar-nordic commented on include/device.h:
```diff
@@ -85,6 +85,26 @@ extern "C" {
 	data, cfg_info, level, prio, NULL)
 
+/**
+ * @def DEVICE_EXEC_FOR_ALL_INSTANCES
```
I think now that #23245 is in it would be cleaner to:

- put this feature into devicetree.h instead of device.h
- make it something like `DT_INST_FOREACH(inst_macro)`, where `inst_macro` just takes 1 argument: the instance number
- `DT_INST_FOREACH` could probably use `DT_NUM_INST(DT_DRV_COMPAT)` to figure out how many instances there are; if not, we can make it possible some other way

io-pa replied:
I changed it.
Now it uses the newly introduced DT_NUM_INST instead of using a while-loop to detect the old-style DT_INST_idx_compat nodes.
Also as suggested now in devicetree.h.

galak commented:
Generally looks go to me.  Will let @mbolivar-nordic and @nordic-krch comment on the UTIL_PASS_ARG_TO_MACRO bits.

gmarull reviewed (commented):
This macro will be really helpful, thanks for this constribution. @io-pa it would be helpful to add some documentation/example.

mbolivar-nordic reviewed (changes requested):
Looks good to me

 I think you could define UTIL_PASS_ARG_TO_MACRO as DT_PASS_ARG_TO_MACRO_INTERNAL at the end of devicetree.h to avoid having to extend the util.h API for this.

Can you please also:

- update doc/guides/dts/howtos.rst to use this in the code example in the "Create struct devices in a driver" section? IMO this should be the standard way to define devices now, whenever possible.
- add a test case for DT_INST_FOREACH to tests/lib/devicetree/src/main.c? Every macro in devicetree.h has at least one test case in that file. One idea would be to pass TEST_GPIO_INIT to DT_INST_FOREACH instead of the way it's done now.

mbolivar-nordic commented:
Might want to change the commit log's "device.h" to "devicetree.h" if you get a chance also. Not a blocker.

io-pa committed:
doc: dts: update devicetree howto to use DT_INST_FOREACH

Updated howto "create struct devices in a driver" section to use
DT_INST_FOREACH instead of manual per-instance macros.

Signed-off-by: Ioannis Papamanoglou <iopapamanoglou@gmail.com>

io-pa committed:
tests: devicetree: use DT_INST_FOREACH

Replace manual per-instance macros with DT_INST_FOREACH
for TEST_GPIO_INIT test.

Signed-off-by: Ioannis Papamanoglou <iopapamanoglou@gmail.com>

io-pa commented:
I changed the helper macro to add `;` at the end, since it's now specific to DT_INST_FOREACH.
That way it supports also functions next to macros and macros that need to be `;`-terminated.

mbolivar-nordic reviewed (approved):
Thank you!

mbolivar-nordic commented on include/devicetree.h:
```diff
@@ -1316,6 +1316,22 @@
+ * Has to accept instance_number as only parameter.
+ */
+#define DT_INST_FOREACH(inst_expr) \
+	UTIL_LISTIFY(DT_NUM_INST(DT_DRV_COMPAT), DT_CALL_WITH_ARG, inst_expr)
```
Just checking -- this does the right thing if DT_NUM_INST(DT_DRV_COMPAT) is 0, right?

io-pa replied:
Uses LISTIFY internally, and that uses REPEAT and REPEAT checks wether the variable is 0 in the beginning so should be fine. I just tested it though to be sure and it looks good.

mbolivar-nordic replied:
Thanks for confirming.

mbolivar-nordic commented:
Shippable failure looks ignorable.

galak commented:
ignoring the shippable failure as its not related.
